### ISSUE TITLE
bigip_virtual_server module is unable to create vip with fasthttp profile

### ISSUE THREAD

**Issue by dbeckham** [F5Networks/f5-ansible#34]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

bigip_virtual_server
##### ANSIBLE VERSION

```
ansible 2.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### PYTHON VERSION

```
Python 2.7.12
```
##### F5 BIGIP LTM VERSION

```
Sys::Version
Main Package
  Product     BIG-IP
  Version     12.1.1
  Build       0.0.184
  Edition     Final
  Date        Thu Aug 11 17:09:01 PDT 2016
```
##### CONFIGURATION

```
retry_files_enabled = False
host_key_checking=False
```
##### OS / ENVIRONMENT

Alpine 3.4 (Docker Container)
##### SUMMARY

The `bigip_virtual_server` module is unable to create a virtual server with the profile "fasthttp". The following error is returned:

```
01070734:3: Configuration error: Found disallowed profile on /Common/myvip: Not Any Of (FastL4 Profile, FastHTTP Profile)
```
##### STEPS TO REPRODUCE

Create a task that uses the `bigip_virtual_server` module with the following configuration parameters:

```
- name: Create virtual server
  bigip_virtual_server:
    name: "myvip"
    destination: "1.1.1.1"
    port: "80"
    pool: "mypool"
    snat: "Automap"
    description: "Primary VIP"
    all_profiles:
      - fasthttp
    password: "{{ bigip_password }}"
    server: "{{ inventory_hostname }}"
    state: "enabled"
    user: "{{ bigip_username }}"
    validate_certs: "{{ validate_certs }}"
```
##### EXPECTED RESULTS

The `bigip_virtual_server` module should successfully create a virtual server with the profile `fasthttp`.

The following is an example tmsh configuration of a Performance (HTTP) virtual server created through the GUI:

```
# tmsh list /ltm virtual
ltm virtual myvip {
    destination 1.1.1.1:http
    ip-protocol tcp
    mask 255.255.255.255
    pool mypool
    profiles {
        fasthttp { }
    }
    source 0.0.0.0/0
    translate-address enabled
    translate-port enabled
    vs-index 10
}
```
##### ACTUAL RESULTS

The following error occurs when the playbook task is run:

```
Using module file /playbooks/library/bigip_virtual_server.py
<lbl11.example.com> ESTABLISH LOCAL CONNECTION FOR USER: root
<lbl11.example.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475100826.05-161604525418252 `" && echo ansible-tmp-1475100826.05-161604525418252="` echo $HOME/.ansible/tmp/ansible-tmp-1475100826.05-161604525418252 `" ) && sleep 0'
<lbl11.example.com> PUT /tmp/tmpx4q3V8 TO /root/.ansible/tmp/ansible-tmp-1475100826.05-161604525418252/bigip_virtual_server.py
<lbl11.example.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1475100826.05-161604525418252/ /root/.ansible/tmp/ansible-tmp-1475100826.05-161604525418252/bigip_virtual_server.py && sleep 0'
<lbl11.example.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1475100826.05-161604525418252/bigip_virtual_server.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1475100826.05-161604525418252/" > /dev/null 2>&1 && sleep 0'
failed: [lbl11.example.com] (item={u'description': u'Primary VIP', u'rules': None, u'destination': u'1.1.1.1', u'profiles': [u'fasthttp'], u'state': u'enabled', u'snat': u'Automap', u'port': 80, u'pool': u'mypool', u'name': u'myvip'}) => {
    "failed": true,
    "invocation": {
        "module_args": {
            "all_enabled_vlans": null,
            "all_profiles": [
                "fasthttp"
            ],
            "all_rules": [
                ""
            ],
            "default_persistence_profile": null,
            "description": "Primary VIP",
            "destination": "1.1.1.1",
            "name": "myvip",
            "partition": "Common",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "pool": "mypool",
            "port": 80,
            "server": "lbl11.example.com",
            "server_port": 443,
            "snat": "Automap",
            "state": "enabled",
            "user": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "validate_certs": false
        },
        "module_name": "bigip_virtual_server"
    },
    "item": {
        "description": "Primary VIP",
        "destination": "1.1.1.1",
        "name": "myvip",
        "pool": "mypool",
        "port": 80,
        "profiles": [
            "fasthttp"
        ],
        "rules": null,
        "snat": "Automap",
        "state": "enabled"
    },
    "msg": "received exception: Error on setting profiles : Server raised fault: 'Exception caught in LocalLB::urn:iControl:LocalLB/VirtualServer::add_profile()\nException: Common::OperationFailed\n\tprimary_error_code   : 17237812 (0x01070734)\n\tsecondary_error_code : 0\n\terror_string         : 01070734:3: Configuration error: Found disallowed profile on /Common/myvip: Not Any Of (FastL4 Profile, FastHTTP Profile)'"
}
        to retry, use: --limit @/playbooks/ltm_application.retry

```

**Comment by caphrim007** [comment]
I believe this is being caused by this module, by default, creating VS's with a tcp profile, and then never removing this tcp profile for the fasthttp profile to be inserted

**Comment by dbeckham** [comment]
Yes, I see that done in the `vs_create()` method before any other work happens in `main()`.

If it helps, I just confirmed that `tcp` and `fasthttp` definitely conflict on my local bigip installation. I edited a regular tcp pool I created with the tmsh command `(tmos) # edit /ltm virtual myvip` to create the following tmsh modify command:

```
modify virtual myvip {
    destination 1.1.1.1:http
    ip-protocol tcp
    mask 255.255.255.255
    pool mypool
    profiles replace-all-with {
        tcp { }
        fasthttp { }
    }
    source 0.0.0.0/0
    translate-address enabled
    translate-port enabled
}
```

Attempting to save and apply the above change gives the same error, where saving with fasthttp profile only, or tcp only successfully modifies the virtual server.

```
01070734:3: Configuration error: Found disallowed profile on /Common/myvip: Not Any Of (FastL4 Profile, FastHTTP Profile)
```

**Comment by caphrim007** [comment]
I can confirm that.

Also, transactions should allow this to work. The update is done in a transaction but the  create is not. Additionally, there is code in there to prevent the tcp profile from being removed. That will need to change as well

**Comment by dbeckham** [comment]
Just some quick feedback on this... I understand why the virtual server module adds tcp by default, but I'm wondering of it's trying to be *too* helpful. Would it be better to push some of the  configuration knowledge back to the user?  For example, if no profile is specified, it would default to tcp, but when any profile is specified, it would require that it be a complete and correct list?  e.g. `['tcp', 'http']`, `['tcp', 'http', 'oneconnect']`, `['fasthttp']`, etc.

This would protect the module against against current issues (Perf Layer 4, Perf HTTP), as well as future profile restrictions in newer versions.

**Comment by caphrim007** [comment]
@dbeckham i've pushed a change that I believe addresses this. The TCP profile was being added because the create method of the SOAP API requires that _some_ profile be supplied. The change takes the profiles that the user supplies and provides those. If the user doesnt provide any, then it defaults to TCP.

Lemme know if it works for you as expected

**Comment by caphrim007** [comment]
I believe this issue has been recently fixed in the repo here. If you can check the updated code to make sure that it addresses the issue you were seeing, I'd appreciate it.

If the problem is still apparent, please file a new issue

**Comment by dbeckham** [comment]
Thanks for fixing this issue!  Due to iRules being unable to be used with the virtual server module (issue #44 and #32), we ended up implementing a workaround and are maintaining virtual servers with the `bigip_command` module instead. It's impossible to see what changes will actually be made when doing it that way, so we are definitely planning on switching back to the virtual server module when it's working again. I'm not sure if we can easily confirm this is fixed at the moment, but will put it through heavy testing again when irules are working.

Thanks much.
